File: Algorithm/BackTracking/주사위 고르기.swift
import Foundation

func solution(_ dice:[[Int]]) -> [Int] {
    var answer: [Int] = []
    var maxWincase: Int = -1
    
    for selection in combination(dice.count / 2) {
        let myCase = dice.enumerated().filter{ selection.contains($0.0) }.map{ $0.1 }
        let enemyCase = dice.enumerated().filter{ selection.contains($0.0) == false }.map{ $0.1 }
                
        let myCurrent = rollDice(myCase, dice.count / 2)
        let enemyCurrent = rollDice(enemyCase, dice.count / 2).sorted()
        let winCase = myCurrent.reduce(0) { $0 + lowerBound(enemyCurrent, $1)}
        if maxWincase < winCase {
            maxWincase = winCase
            answer = selection.map { $0 + 1}
        }
        
    }
    
    return answer

} // solution

func combination(_ r: Int) -> [[Int]] {
    var result: [[Int]] = []
        
    func dfs(_ com: [Int], _ cur: Int, _ depth: Int) {
        if depth == r {
            result.append(com)
            return
        }   
        
        for i in cur..<(r * 2) {
            dfs(com + [i], i + 1, depth + 1)
        }
        return
    } // dfs
    
    dfs([], 0, 0)
        
    return result
} // combination

func rollDice(_ dice: [[Int]], _ diceCount:Int) -> [Int] {
    var result: [Int] = []
    
    func dfs(_ sum: Int, _ depth: Int) {
        if depth == diceCount {
            result.append(sum)
            return
        }
        
        for i in 0..<6 {
            dfs(sum + dice[depth][i], depth + 1)
        }
        
        return
    }
    
    dfs(0, 0)
    return result
} // rollDice

func lowerBound(_ arr: [Int], _ target: Int) -> Int {
    var start: Int = 0, end: Int = arr.count
    
    while start < end {
        let mid = (start + end) / 2
        
        if arr[mid] < target {
            start = mid + 1
        } else {
            end = mid
        }
    }
    
    return start
}

File: Algorithm/BackTracking/주사위 고르기.py
from itertools import combinations
import bisect

def solution(dice):
    answer = []
    dice_type = [i for i in range(len(dice))]
    temp_score = []
    max_win = -1
    
    def dfs(dice_idx_list, tot, depth):
        if depth == len(dice_idx_list):
            temp_score.append(tot)
            return
    
        for number in dice[dice_idx_list[depth]]:
            dfs(dice_idx_list, tot + number, depth + 1)
    ## dfs
    
    for my in combinations(dice_type, len(dice) // 2):
        enemy = [i for i in range(len(dice)) if not i in my]
        #print(my, enemy)
        
        temp_score = []
        dfs(my, 0, 0)
        my_score = temp_score[:]
        
        temp_score = []
        dfs(enemy, 0, 0)
        enemy_score = sorted(temp_score[:])
        
        win = 0
        for target in my_score:
            win += bisect.bisect_left(enemy_score, target)
        
        #print(my, win)
        if max_win < win:
            max_win = win
            answer = [m+1 for m in my]
    
    return answer
## solution